library/std/src/sys/unix/time.rs

```diff
@@ -233,31 +232,46 @@ mod inner {
     }
 
     fn info() -> mach_timebase_info {
-        static mut INFO: mach_timebase_info = mach_timebase_info { numer: 0, denom: 0 };
-        static STATE: AtomicUsize = AtomicUsize::new(0);
+        static INITIALIZED: AtomicBool = AtomicBool::new(false);
+        static INFO_BITS: AtomicU64 = AtomicU64::new(0);
+
+        // If a previous thread has filled in this global INITIALIZED, use that.
+        if INITIALIZED.load(Ordering::Acquire) {
+            // The Acquire/Release pair used for INITIALIZED ensures that this
+            // load can see the corresponding `INFO_BITS` store, despite them
+            // both being Relaxed.
+            return info_from_bits(INFO_BITS.load(Ordering::Relaxed));
+        }
+
+        // ... otherwise learn for ourselves ...
+        extern "C" {
+            fn mach_timebase_info(info: mach_timebase_info_t) -> kern_return_t;
+        }
 
+        let mut info = info_from_bits(0);
         unsafe {
-            // If a previous thread has filled in this global state, use that.
-            if STATE.load(SeqCst) == 2 {
-                return INFO;
-            }
+            mach_timebase_info(&mut info);
+        }
 
-            // ... otherwise learn for ourselves ...
-            let mut info = mem::zeroed();
-            extern "C" {
-                fn mach_timebase_info(info: mach_timebase_info_t) -> kern_return_t;
-            }
+        // Note: This is racy, but the race is against other threads trying to
+        // write the same value.
```

Comment:
The only way this isn't true is if mach_timebase_info fails for one thread and not for the others.

I don't think in practice this function is fallible (and if it is I suspect it will fail for all the threads, although I have no real evidence of this).

Anyway, it's a correctness concern and not soundness, and the old code shared that correctness concern since the thread that acquires the lock is arbitrary and could be one that failed, so this patch doesn't meaningfully change the situation here.